<a href="https://colab.research.google.com/github/heispv/python-programming-july2023/blob/master/03_python_programming.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## immutablitiy

object i hast ke gheire ghabele taghir mibashad.

In [ ]:
my_string = 'hello'
my_string[1] = 'a'

TypeError: ignored

> reshte ha gheire ghabele taghir hastan va yek item az oonaro be komake indexing nemitoonim taghir bedim, agar khastim yek chizi ro dakhelesh taghir bedim bayad darkol oon variable ro reassign bokonim.

In [ ]:
my_string = 'hallo'

In [ ]:
my_string

'hallo'

## .capitalize()

miad va baraye ma avalin harfe oon string ro capital letter mikone.

In [ ]:
second_string = my_string.capitalize()

In [ ]:
my_string

'hallo'

In [ ]:
second_string

'Hallo'

## .casefold()

miad va baraye ma kolle string ro ba lower case minvise.

In [ ]:
ex_2 = 'MYFRIENDS'
ex_2.casefold()

'myfriends'

## .count(value, [start], [stop])

miad va oon string i ke dakhele parantez moshakhas kardin ro mige ke chan bar tekrar shode.

In [ ]:
ex_3 = 'aabbcccb'
ex_3.count('c')

3

> mitoonim biaim va az vurudi haye optional e method e count estefade konim, avvalin vurudi optional (dovomin vurudi kolli) e oon start va dovomish end index hastesh.

In [ ]:
ex_3.count('b', 3, 7)

1

## .find(value, [start], [stop])

find miad va avvalin value i ke peyda mikone ro behet index ro mide, va baghiaro engar nemibine. agar khastim ke oon baghiaro behemoon namayesh bede bayad az start va stop estefade konim.

agar too oon bazei ke moshakhas kardim chizi peyda nakard be ma -1 ro khuruji mide.

In [ ]:
ex_4 = 'artifical intelligence'
ex_4.find('i', 6, 10)

-1

## .replace(oldvalue, newvalue, [count])

miad va yek value ro ba value dge jaygozin mikone.


In [ ]:
ex_5 = 'aax123334'
ex_5_1 = ex_5.replace('3', 'y')

In [ ]:
ex_5_1

'aax12yyy4'

In [ ]:
ex_6 = '11111111'
ex_6.replace('1', 'a', 3)

'aaa11111'

## .split([separator], [maxsplit])

miaim va yek reshte ro moshakhas mikonim va oon bar asase oon reshte miad va reshtamoon ro split mikone. maxsplit dar halate addi maximum ama agar bekhaim mitoonim baraye oon ham meghdar moshakhas bokonim.

In [ ]:
txt = "welcome to the jungle"

x = txt.split()

print(x)

['welcome', 'to', 'the', 'jungle']


In [ ]:
txt = "welcome to the jungle"

x = txt.split(' ', 2)

print(x)

['welcome', 'to', 'the jungle']


> baghiyeye method haro mitoonim az [w3school](https://www.w3schools.com/) bekhoonin.

# Boolean

boolean ha ya True hastna ya False hastan.

True/False

1/0

presence/absence

In [ ]:
bool(0)

False

In [ ]:
bool(-1.3)

True

In [ ]:
bool(10)

True

In [ ]:
type(bool([])) #list

bool

In [ ]:
bool([1]) #list

True

In [ ]:
bool(()) #tuple

False

In [ ]:
bool((1, 2))

True

## example

yek barname benvisid ke sale tarikh tavalod ro begire va sene shoma ro mohasebe kone.

In [ ]:
tarikh_tavalod = input('Sen khod ra vared konid: ')
type(tarikh_tavalod), tarikh_tavalod

Sen khod ra vared konid: 1377


(str, '1377')

> deghat konim ke aln 1377 be onave reshte zakhire shode.

In [ ]:
# input changed to int
tarikh_tavalod = int(input('Sen khod ra vared konid: '))

sen = 1402 - tarikh_tavalod

print(f'Sen e shoma {sen} ast')

Sen khod ra vared konid: 1377
Sen e shoma 25 ast


## List

listi az object haye mokhtalef hastand ke mitoonin oonaro zakhire konin va be komake indexing oonaro estekhraj konin.

In [ ]:
list_example = [1, 'sib', {'name': 'john'}, (1, 2, 3)]

In [ ]:
list_example[2]

{'name': 'john'}

In [ ]:
mive = ['hulu', 'sib', 'portoghal','kivi']

In [ ]:
mive, type(mive)

(['hulu', 'sib', 'portoghal', 'kivi'], list)

In [ ]:
len(mive)

4

In [ ]:
mive[1:4]

['sib', 'portoghal', 'kivi']

In [ ]:
mive[1:4:2]

['sib', 'kivi']

In [ ]:
mive[-1]

'kivi'

In [ ]:
mive

['hulu', 'sib', 'portoghal', 'kivi']

In [ ]:
mive[1] = 'sib_ghermez'

In [ ]:
mive

['hulu', 'sib_ghermez', 'portoghal', 'kivi']

> mikham character z ro dakhele sib_ghermez begiram

In [ ]:
mive[1][-1]

'z'

In [ ]:
mive_2 = mive

In [ ]:
mive

['hulu', 'sib_ghermez', 'portoghal', 'kivi']

In [ ]:
mive_2

['hulu', 'sib_ghermez', 'portoghal', 'kivi']

In [ ]:
mive_2[1] = 'sib_sabz'

In [ ]:
mive_2

['hulu', 'sib_sabz', 'portoghal', 'kivi']

In [ ]:
mive

['hulu', 'sib_sabz', 'portoghal', 'kivi']

> agar khastim yek list ro taghir bedim va oon yeki taghir nakone bayad be komake `.copy()` in karo anjam bedim.

In [ ]:
mive_2 = mive.copy()

In [ ]:
mive_2

['hulu', 'sib_sabz', 'portoghal', 'kivi']

In [ ]:
mive_2[0] = 'sib_ghermez'

In [ ]:
mive_2

['sib_ghermez', 'sib_sabz', 'portoghal', 'kivi']

In [ ]:
mive

['hulu', 'sib_sabz', 'portoghal', 'kivi']

## .append(object)

miad va be akhare list e ma yek object ro ezafe mikone.

In [ ]:
ex_list = [1, 2, 3]

In [ ]:
ex_list.append('a')

In [ ]:
ex_list

[1, 2, 3, 'a']

In [ ]:
ex_list.append(['b', 'c', 'd'])
ex_list

[1, 2, 3, 'a', ['b', 'c', 'd']]

In [ ]:
len(ex_list)

5

## .extend()

In [ ]:
ex_list_2 = [1, 2]

In [ ]:
ex_list_2.extend(['b', 'c', 'd'])
ex_list_2

[1, 2, 'b', 'c', 'd']